src/loaders/gtdb/compute_genome_attribs.py

```diff
@@ -0,0 +1,369 @@
+                      os.path.isdir(os.path.join(source_data_dir, path))]
+
+    if not all_genome_ids:
+        raise ValueError('Please download genome files first.')
+
+    # get genome ids for the running job
+    if genome_id_file:
+        # parse genome ids from tab separated file
+        with genome_id_file:
+            df = pd.read_csv(genome_id_file, sep='\t')
+            try:
+                genome_ids = df['genome_id']
+            except KeyError:
+                raise ValueError('Please ensure genome_id column exists in the genome id file.')
+
+        if not set(all_genome_ids) >= set(genome_ids):
+            raise ValueError(f'Genome files for {set(genome_ids) - set(all_genome_ids)} are not downloaded yet')
+    else:
+        # executing all genomes
+        genome_ids = all_genome_ids
+
+    total_count = len(genome_ids)
+
+    if not threads:
+        threads = max(int(multiprocessing.cpu_count() * min(SYSTEM_UTILIZATION, 1)), 1)
+    threads = max(1, threads)
+
+    for tool in tools:
+        # execute each tooling in series
+        try:
+            comp_ops = getattr(sys.modules[__name__], tool)
+        except AttributeError as e:
+            raise ValueError(
+                f'Please implement method for: [{tool}]\n'
+                f'NOTE: Method name should be exactly the same as the tool name') from e
+
+        # place computed results to COLLECTION_DATA_DIR directory
+        work_dir = os.path.join(root_dir, COLLECTION_DATA_DIR, kbase_collection,
+                                load_ver, tool)
+
+        start = time.time()
+        print(f"Start executing {tool} with {threads} threads")
+        if tool in SERIES_TOOLS:
+            failed_ids = comp_ops(genome_ids, work_dir, source_data_dir, debug, threads)
+        else:
+            # call tool execution in parallel
+            batch_size = max(math.floor(threads / program_threads), 1)
+            chunk_size = math.ceil(len(genome_ids) / batch_size)  # distribute genome ids evenly across batches
+            batch_input = [(genome_ids[i: i + chunk_size],
+                            work_dir,
+                            source_data_dir,
+                            debug,
+                            program_threads,
+                            batch_number,
+                            node_id) for batch_number, i in enumerate(range(0, total_count, chunk_size))]
+            pool = multiprocessing.Pool(processes=threads)
+            batch_result = pool.starmap(comp_ops, batch_input)
+            failed_ids = list(itertools.chain.from_iterable(batch_result))
+        print(
+            f'In total used {round((time.time() - start) / 60, 2)} minutes to execute {tool} for {total_count} genomes')
+
+        if failed_ids:
+            print(f'Failed to execute {tool} for {failed_ids}')
+        else:
+            print(f'Successfully executed all genomes for {tool}')
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
At some point the code should be able to pick up where it left off, or tell which runs failed and rerun just those, but that's later work

Reply:
👍  I haven't thought through this. That's also related to checkpointing.